src/data/list/basic.lean

```diff
@@ -1941,6 +1941,50 @@ begin
     apply hl _ (list.mem_cons_self _ _) }
 end
 
+/-- Transform a `l : list α` folded via `foldr` across a motive that holds
+for the seed element `b : β` and for all incremental `op : α → β → β`
+performed on the elements `(a : α) ∈ l`. -/
+def foldr_rec_on {C : β → Sort*} (l : list α) (op : α → β → β) (b : β) (hb : C b)
+  (hl : ∀ (b : β) (hb : C b) (a : α) (ha : a ∈ l), C (op a b)) :
+  C (foldr op b l) :=
+begin
+  induction l with hd tl IH,
+  { exact hb },
+  { refine hl _ _ hd (mem_cons_self hd tl),
+    refine IH _,
+    intros y hy x hx,
+    exact hl y hy x (mem_cons_of_mem hd hx) }
+end
+
+/-- Transform a `l : list α` folded via `foldl` across a motive that holds
+for the seed element `b : β` and for all incremental `op : β → α → β`
+performed on the elements `(a : α) ∈ l`. -/
+def foldl_rec_on {C : β → Sort*} (l : list α) (op : β → α → β) (b : β) (hb : C b)
+  (hl : ∀ (b : β) (hb : C b) (a : α) (ha : a ∈ l), C (op b a)) :
+  C (foldl op b l) :=
+begin
+  induction l with hd tl IH generalizing b,
+  { exact hb },
+  { refine IH _ _ _,
+    { intros y hy x hx,
+      exact hl y hy x (mem_cons_of_mem hd hx) },
+    { exact hl b hb hd (mem_cons_self hd tl) } }
+end
+
+@[simp] lemma foldr_rec_on_nil {C : β → Sort*} (op : α → β → β) (b) (hb : C b) (hl) :
+  foldr_rec_on [] op b hb hl = hb := rfl
+
+@[simp] lemma foldr_rec_on_cons {C : β → Sort*} (x : α) (l : list α)
+  (op : α → β → β) (b) (hb : C b)
+  (hl : ∀ (b : β) (hb : C b) (a : α) (ha : a ∈ (x :: l)), C (op a b)) :
+  foldr_rec_on (x :: l) op b hb hl = hl _ (foldr_rec_on l op b hb
+    (λ b hb a ha, hl b hb a (mem_cons_of_mem _ ha))) x (mem_cons_self _ _) := rfl
+
+@[simp] lemma foldl_rec_on_nil {C : β → Sort*} (op : β → α → β) (b) (hb : C b) (hl) :
+  foldl_rec_on [] op b hb hl = hb := rfl
+
+-- TODO: add foldl_rec_on_cons
```

Comment:
I assume you mean "to do in another PR", as opposed to "I plan to come back to this?

Reply:
Correct. I think in general there aren't a ton of uses of foldl. And I'm having mind block figuring out the right recursive form for the cons proof.

Reply:
I think `foldl_rec_on_cons` is unlikely to have any usable form at all... Maybe just remove the TODO?